### PR TITLE
ci.ocp: Document openshift pipeline and manual bisection

### PR DESCRIPTION
**Pull request by ldoktor**
In https://github.com/kata-containers/kata-containers/pull/9229 we allowed running the pipeline with a custom kata-deploy image, let's go even further and add helper scripts to allow bisection over PRs. While on it also document the kata-containers pipeline.

Further addresses the #9228

### CODE REVIEW

File: ci/openshift-ci/bisect-range.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (GabyCT) [inline ci/openshift-ci/bisect-range.sh:1]: probably add something like this
```
# Copyright (c) 2024 Red Hat, Inc.
#
# SPDX-License-Identifier: Apache-2.0
#
```

File: ci/openshift-ci/sample-test-reproducer.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+# Sample script to deploy, configure, run E2E_TEST and soft-cleanup
```
Comment (GabyCT) [inline ci/openshift-ci/sample-test-reproducer.sh:2]: the same comment as before

**Comment by ldoktor** [comment]
Changes:

* Added license to scripts

**Review by GabyCT — APPROVED** [review]

File: ci/openshift-ci/cleanup.sh [hunk 3]
```diff
@@ -25,6 +25,10 @@ WORKAROUND_9206_CRIO=${WORKAROUND_9206_CRIO:-no}
 # Ignore errors as we want best-effort-approach here
 trap - ERR
 
+# Delete webhook resources
```
Comment (wainersm) [inline ci/openshift-ci/cleanup.sh:28]: Checked that it reverts the resources installed in https://github.com/kata-containers/kata-containers/blob/main/ci/openshift-ci/cluster/deploy_webhook.sh

File: ci/openshift-ci/README.md [hunk 4]
```diff
@@ -8,3 +8,142 @@ There are 2 pipelines, history and logs can be accessed here:
...
+#!/bin/bash -e
+# Setup your kubectl and check it's accessible by
+kubectl nodes
+# Deploy kata (set KATA_DEPLOY_IMAGE to override the default kata-deploy-ci:latest image)
+./test.sh
+# Deploy the webhook
+KATA_RUNTIME=kata-qemu cluster/deploy_webhook.sh
+```
+
+This should ensure kata-containers as well as kata-webhook are installed and
```
Comment (wainersm) [inline ci/openshift-ci/README.md:29]: hi @ldoktor ! Does it sounds obvious that you should wait the kata pods and webhook be up and running? Or maybe it's worth a note.

File: ci/openshift-ci/README.md [hunk 5]
```diff
@@ -8,3 +8,142 @@ There are 2 pipelines, history and logs can be accessed here:
...
+* what should be executed (perhaps running the setup is enough for you or
+  you might want to be looking for specific failures...)
+* use ``timeout`` to interrupt execution in case you know things should be faster
+
+Executing that script with the GOOD commit should pass
+``./sample-test-reproducer.sh quay.io/kata-containers/kata-deploy-ci:kata-containers-d7afd31fd40e37a675b25c53618904ab57e74ccd-amd64``
+and fail when executed with the BAD commit
+``./sample-test-reproducer.sh quay.io/kata-containers/kata-deploy-ci:kata-containers-9f512c016e75599a4a921bd84ea47559fe610057-amd64``.
+
+To get the list of all tags in between those two PRs you can use the
```
Comment (wainersm) [inline ci/openshift-ci/README.md:121]: Worth mention it requires `skopeo` installed? I think this tool isn't installed by default on most distros

**Comment by wainersm** [comment]
Labeled with `force-skip-ci` because this PR won't be tested by functional tests, so let's avoid waste.

File: ci/openshift-ci/README.md (hunk 4 above)
Comment (ldoktor) [reply ci/openshift-ci/README.md:29]: Both the "test.sh" and "deploy_webhook.sh" ensures it's up and running so no need to double-check

File: ci/openshift-ci/README.md (hunk 5 above)
Comment (ldoktor) [reply ci/openshift-ci/README.md:121]: We are not mentioning the other tools, I'd assume people will get the error message... Anyway if you insist I can add it.

Comment (wainersm) [reply ci/openshift-ci/README.md:121]: It's your call. I don't have a strong opinion ;)

**Review by wainersm — APPROVED** [review]
LGTM @ldoktor !

I'd love to see the bisecter being used on kata outside of openshift-ci context.

**Comment by ldoktor** [comment]
> LGTM @ldoktor !
> 
> I'd love to see the bisecter being used on kata outside of openshift-ci context.

Bisecter is cool, if only I had time to add the git support for it... (currently one needs to extract the shas and feed it to bisecter)

File: ci/openshift-ci/README.md [hunk 6]
```diff
@@ -8,3 +8,142 @@ There are 2 pipelines, history and logs can be accessed here:
...
+``bisect-range.sh`` script
+
+```bash
+./bisect-range.sh d7afd31fd40e37a675b25c53618904ab57e74ccd 9f512c016e75599a4a921bd84ea47559fe610057
+```
+
+[!NOTE]
+The tagged images are only built per PR, not for individual commits. See
+[kata-deploy-ci](quay.io/kata-containers/kata-deploy-ci) to see the
```
Comment (wainersm) [inline ci/openshift-ci/README.md:130]: Hi @ldoktor !

Static-check failing and I think it fails to follow that link:
```
time="2024-04-11T17:20:18.393783456Z" level=fatal msg="found 1 parse error:\nstat ci/openshift-ci/quay.io/kata-containers/kata-deploy-ci: no such file or directory" commit=d37b58bfc9fa0caf57db1c73c3bf37304fc1bf79 name=kata-check-markdown pid=4996 source=check-markdown version=0.0.1
```

Comment (ldoktor) [reply ci/openshift-ci/README.md:130]: Oh, good catch. Let me prefix `https://` then.

**Comment by wainersm** [comment]
Hi @ldoktor ! This time static-check complains about misspelled words /o\

```
2024-05-14T14:34:06.6518315Z [kata-spell-check.sh:125] INFO: Spell checking file 'README.md'
2024-05-14T14:34:06.7885026Z [kata-spell-check.sh:286] INFO: Spell check successful for file: 'README.md'
2024-05-14T14:34:06.7997963Z [kata-spell-check.sh:125] INFO: Spell checking file 'ci/openshift-ci/README.md'
2024-05-14T14:34:07.3541674Z [kata-spell-check.sh:275] WARNING: Word 'MonitorTest': did you mean one of the following?: Monitor Test, Monitor-test, Mosterton, Testerton, Morriston, Misterton
2024-05-14T14:34:07.3574335Z [kata-spell-check.sh:275] WARNING: Word 'bisecter': did you mean one of the following?: bisector, bisected, bisect er, bisect-er, bisect
2024-05-14T14:34:07.3607636Z [kata-spell-check.sh:275] WARNING: Word 'bisecter': did you mean one of the following?: bisector, bisected, bisect er, bisect-er, bisect
2024-05-14T14:34:07.3641549Z [kata-spell-check.sh:275] WARNING: Word 'bisecter': did you mean one of the following?: bisector, bisected, bisect er, bisect-er, bisect
2024-05-14T14:34:07.3676374Z [kata-spell-check.sh:275] WARNING: Word 'ci': did you mean one of the following?: OCI
2024-05-14T14:34:07.3709816Z [kata-spell-check.sh:275] WARNING: Word 'ci': did you mean one of the following?: OCI
2024-05-14T14:34:07.3744286Z [kata-spell-check.sh:275] WARNING: Word 'clonned': did you mean one of the following?: cloned, conned, clonked, clowned, clone
2024-05-14T14:34:07.3778290Z [kata-spell-check.sh:275] WARNING: Word 'everythings': did you mean one of the following?: everything, everything's, every things, every-things, everything s, everlastings
2024-05-14T14:34:07.3812002Z [kata-spell-check.sh:275] WARNING: Word 'io': did you mean one of the following?: oi, Io, ii, i, o, ion, bio, Rio, is, so, in, no, it, to, lo
2024-05-14T14:34:07.3845844Z [kata-spell-check.sh:275] WARNING: Word 'io': did you mean one of the following?: oi, ii, i, o, ion, bio, Rio, Pio, is, so, in, no, it, to, lo
2024-05-14T14:34:07.3878735Z [kata-spell-check.sh:275] WARNING: Word 'openshift': did you mean one of the following?: open shift, open-shift, downshift
2024-05-14T14:34:07.3912405Z [kata-spell-check.sh:275] WARNING: Word 'openshift': did you mean one of the following?: open shift, open-shift, downshift
2024-05-14T14:34:07.3945486Z [kata-spell-check.sh:275] WARNING: Word 'openshift': did you mean one of the following?: open shift, open-shift, downshift
2024-05-14T14:34:07.3978191Z [kata-spell-check.sh:275] WARNING: Word 'openshift': did you mean one of the following?: open shift, open-shift, downshift
2024-05-14T14:34:07.4010722Z [kata-spell-check.sh:275] WARNING: Word 'openshift': did you mean one of the following?: open shift, open-shift, downshift
2024-05-14T14:34:07.4044305Z [kata-spell-check.sh:275] WARNING: Word 'openshift': did you mean one of the following?: open shift, open-shift, downshift
2024-05-14T14:34:07.4052988Z [kata-spell-check.sh:284] ERROR: Spell check failed for file: 'ci/openshift-ci/README.md'
2024-05-14T14:34:07.4063886Z [static-checks.sh:899] INFO: spell check failed for document ci/openshift-ci/README.md
```

I think the static-checker's spell-checker on this repository is just a wrapper to https://github.com/kata-containers/tests/tree/main/cmd/check-spelling . So if needed new words to the dictionary then you have to add in the database there.

**Pull request merged**